File: controllers/employeeController.js
// =====================================================
// Controller: Nhân viên (CRUD + Search + Paginate)
// =====================================================

const EmployeeModel = require('../models/employeeModel');
const path = require('path');
const fs = require('fs');

const getAll = async (req, res) => {
    try {
        const options = {
            page: parseInt(req.query.page) || 1,
            limit: parseInt(req.query.limit) || 10,
            search: req.query.search || '',
            maPhongBan: req.query.maPhongBan || '',
            maChucVu: req.query.maChucVu || '',
            trangThai: req.query.trangThai || '',
            sortBy: req.query.sortBy || 'MaNhanVien',
            sortDir: req.query.sortDir || 'ASC'
        };

        const [employees, total] = await Promise.all([
            EmployeeModel.getAll(options),
            EmployeeModel.count({
                search: options.search,
                maPhongBan: options.maPhongBan,
                trangThai: options.trangThai
            })
        ]);

        return res.json({
            success: true,
            data: employees,
            pagination: {
                page: options.page,
                limit: options.limit,
                total,
                totalPages: Math.ceil(total / options.limit)
            }
        });
    } catch (error) {
        console.error('Get employees error:', error);
        return res.status(500).json({ success: false, message: 'Lỗi lấy danh sách nhân viên' });
    }
};

const getOne = async (req, res) => {
    try {
        const employee = await EmployeeModel.findById(req.params.id);
        if (!employee) {
            return res.status(404).json({ success: false, message: 'Không tìm thấy nhân viên' });
        }
        return res.json({ success: true, data: employee });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi máy chủ' });
    }
};

const create = async (req, res) => {
    try {
        const data = req.body;

        // Kiểm tra MaNV đã tồn tại
        if (data.maNV) {
            const exists = await EmployeeModel.maNVExists(data.maNV);
            if (exists) {
                return res.status(400).json({ success: false, message: 'Mã nhân viên đã tồn tại' });
            }
        }

        // Avatar
        if (req.file) {
            data.avatar = `/uploads/avatars/${req.file.filename}`;
        }

        const id = await EmployeeModel.create(data);
        const employee = await EmployeeModel.findById(id);

        return res.status(201).json({
            success: true,
            message: 'Thêm nhân viên thành công',
            data: employee
        });
    } catch (error) {
        console.error('Create employee error:', error);
        return res.status(500).json({ success: false, message: 'Lỗi thêm nhân viên: ' + error.message });
    }
};

const update = async (req, res) => {
    try {
        const { id } = req.params;
        const data = req.body;

        const existing = await EmployeeModel.findById(id);
        if (!existing) {
            return res.status(404).json({ success: false, message: 'Không tìm thấy nhân viên' });
        }

        // Avatar mới
        if (req.file) {
            data.avatar = `/uploads/avatars/${req.file.filename}`;
            // Xóa avatar cũ nếu có
            if (existing.Avatar) {
                const oldPath = path.join(__dirname, '..', existing.Avatar);
                if (fs.existsSync(oldPath)) fs.unlinkSync(oldPath);
            }
        }

        await EmployeeModel.update(id, data);
        const updated = await EmployeeModel.findById(id);

        return res.json({
            success: true,
            message: 'Cập nhật nhân viên thành công',
            data: updated
        });
    } catch (error) {
        console.error('Update employee error:', error);
        return res.status(500).json({ success: false, message: 'Lỗi cập nhật nhân viên' });
    }
};

const remove = async (req, res) => {
    try {
        const existing = await EmployeeModel.findById(req.params.id);
        if (!existing) {
            return res.status(404).json({ success: false, message: 'Không tìm thấy nhân viên' });
        }

        await EmployeeModel.delete(req.params.id);
        return res.json({ success: true, message: 'Xóa nhân viên thành công' });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi xóa nhân viên: ' + error.message });
    }
};

const getSimple = async (req, res) => {
    try {
        const employees = await EmployeeModel.getAllSimple();
        return res.json({ success: true, data: employees });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi máy chủ' });
    }
};

module.exports = { getAll, getOne, create, update, remove, getSimple };

File: quan.comdemo/quan.comdemo/src/main/java/K23/CNT1/_5/DAQ/quan/comdemo/controller/EmployeeController.java
package K23.CNT1._5.DAQ.quan.comdemo.controller;

import K23.CNT1._5.DAQ.quan.comdemo.dto.ApiResponse;
import K23.CNT1._5.DAQ.quan.comdemo.entity.NhanVien;
import K23.CNT1._5.DAQ.quan.comdemo.service.EmployeeService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.File;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@RestController
@RequestMapping("/api/employees")
public class EmployeeController {

    @Autowired
    private EmployeeService employeeService;

    @GetMapping
    public ResponseEntity<ApiResponse<List<NhanVien>>> getAll(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "10") int limit,
            @RequestParam(required = false, defaultValue = "") String search,
            @RequestParam(required = false) Integer maPhongBan,
            @RequestParam(required = false) Integer maChucVu,
            @RequestParam(required = false, defaultValue = "") String trangThai,
            @RequestParam(defaultValue = "MaNhanVien") String sortBy,
            @RequestParam(defaultValue = "ASC") String sortDir) {

        Page<NhanVien> resultPage = employeeService.getAll(page, limit, search, maPhongBan, maChucVu, trangThai, sortBy, sortDir);

        Map<String, Object> pagination = new HashMap<>();
        pagination.put("page", page);
        pagination.put("limit", limit);
        pagination.put("total", resultPage.getTotalElements());
        pagination.put("totalPages", resultPage.getTotalPages());

        return ResponseEntity.ok(ApiResponse.ok(resultPage.getContent(), pagination));
    }

    @GetMapping("/simple")
    public ResponseEntity<ApiResponse<List<NhanVien>>> getSimple() {
        return ResponseEntity.ok(ApiResponse.ok(employeeService.getAllSimple()));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<NhanVien>> getOne(@PathVariable Integer id) {
        NhanVien nv = employeeService.getOne(id);
        if (nv == null) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(ApiResponse.error("Không tìm thấy nhân viên"));
        }
        return ResponseEntity.ok(ApiResponse.ok(nv));
    }

    @PostMapping
    public ResponseEntity<ApiResponse<NhanVien>> create(
            @RequestParam Map<String, Object> params,
            @RequestParam(value = "avatar", required = false) MultipartFile avatarFile) {
        try {
            String avatarPath = handleUpload(avatarFile);
            NhanVien created = employeeService.create(params, avatarPath);
            return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.ok("Thêm nhân viên thành công", created));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage()));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Lỗi thêm nhân viên: " + e.getMessage()));
        }
    }

    @PutMapping("/{id}")
    public ResponseEntity<ApiResponse<NhanVien>> update(
            @PathVariable Integer id,
            @RequestParam Map<String, Object> params,
            @RequestParam(value = "avatar", required = false) MultipartFile avatarFile) {
        try {
            String avatarPath = handleUpload(avatarFile);
            NhanVien updated = employeeService.update(id, params, avatarPath);
            return ResponseEntity.ok(ApiResponse.ok("Cập nhật nhân viên thành công", updated));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Lỗi cập nhật nhân viên: " + e.getMessage()));
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<ApiResponse<String>> delete(@PathVariable Integer id) {
        try {
            employeeService.delete(id);
            return ResponseEntity.ok(ApiResponse.ok("Xóa nhân viên thành công", null));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Lỗi xóa nhân viên: " + e.getMessage()));
        }
    }

    private String handleUpload(MultipartFile avatarFile) throws Exception {
        if (avatarFile != null && !avatarFile.isEmpty()) {
            String fileName = UUID.randomUUID().toString() + "_" + avatarFile.getOriginalFilename();
            File uploadDir = new File("uploads/avatars");
            if (!uploadDir.exists()) uploadDir.mkdirs();
            File dest = new File(uploadDir, fileName);
            avatarFile.transferTo(dest);
            return "/uploads/avatars/" + fileName;
        }
        return null;
    }
}
